clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3130,26 +3130,53 @@ static void handleWorkGroupSize(Sema &S, Decl *D, const ParsedAttr &AL) {
       return;
   }
 
-  if (WorkGroupAttr *ExistingAttr = D->getAttr<WorkGroupAttr>()) {
-    ASTContext &Ctx = S.getASTContext();
-    Optional<llvm::APSInt> XDimVal = XDimExpr->getIntegerConstantExpr(Ctx);
-    Optional<llvm::APSInt> YDimVal = YDimExpr->getIntegerConstantExpr(Ctx);
-    Optional<llvm::APSInt> ZDimVal = ZDimExpr->getIntegerConstantExpr(Ctx);
-    Optional<llvm::APSInt> ExistingXDimVal = ExistingAttr->getXDimVal(Ctx);
-    Optional<llvm::APSInt> ExistingYDimVal = ExistingAttr->getYDimVal(Ctx);
-    Optional<llvm::APSInt> ExistingZDimVal = ExistingAttr->getZDimVal(Ctx);
-
-    // Compare attribute arguments value and warn for a mismatch.
-    if (ExistingXDimVal != XDimVal || ExistingYDimVal != YDimVal ||
-        ExistingZDimVal != ZDimVal) {
-      S.Diag(AL.getLoc(), diag::warn_duplicate_attribute) << AL;
-      S.Diag(ExistingAttr->getLocation(), diag::note_conflicting_attribute);
+  ASTContext &Ctx = S.getASTContext();
+
+  if (!XDimExpr->isValueDependent() && !YDimExpr->isValueDependent() &&
```

Comment:
Yes addIntelTripleArgAttr() calls verifyIntegerConstantExpression () where the X, Y, and Z dim expressions in this case are obtained from A->getXDim(), A->getYDim(), and A->getZDim()

The X, Y, and Z dim expressions in this case here are obtained from AL and here we are checking that those expressions are value-dependent before calling VerifyIntegerConstantExpression() on them here. Otherwise this will cause a failed assertion on a dependent value. This does not go to to the addIntelTripleArgAttr() function.

We need to do a bot of code of duplication to handle test cases below otherwise assertion will happen if any of the argument expression is not a constant integer expression

struct TRIFuncObjBad15 {
[[intel::num_simd_work_items(3]]
[[intel::reqd_work_group_size(3.f)]] // expected-error{{integral constant expression must have integral or unscoped enumeration type, not 'float'}}
void operator()() const {}
};

struct TRIFuncObjBad17 {
[[intel::num_simd_work_items(3)]]
[[intel::reqd_work_group_size(3, 3, 3.f)]] // expected-error{{integral constant expression must have integral or unscoped enumeration type, not 'float'}}
void operator()() const {}
};
if (XDim.isInvalid())
      return;
    XDimExpr = XDim.get();

    if (YDim.isInvalid())
      return;
    YDimExpr = YDim.get();

    if (ZDim.isInvalid())
      return;
    ZDimExpr = ZDim.get();

Reply:
>>I'm not sure if anything can/should be done in this PR but this is something we probably should look into when making the attribute more upstream friendly.


I will refactor both these attributes in a separate patch to be more upstream friendly.

Reply:
Hmm I see. What happens if we don't call `addIntelTripleArgAttr` in this function and just add the attribute directly in this function? If the checks in  `addIntelTripleArgAttr `are basically for handling `NumSimdWorkItemsAttr`, I'm not sure I understand why it should be called in this function.

Reply:
>>What happens if we don't call addIntelTripleArgAttr in this function and just add the attribute directly in this function?

Template support will not work here

Reply:
> Hmm I see. What happens if we don't call `addIntelTripleArgAttr` in this function and just add the attribute directly in this function? If the checks in `addIntelTripleArgAttr `are basically for handling `NumSimdWorkItemsAttr`, I'm not sure I understand why it should be called in this function.

Sorry i probably confused you. 

Inside handleNumSimdWorkItemsAttr,(), i am handling this test case.


struct TRIFuncObjBad {
[[intel::reqd_work_group_size(4)]] 
[[intel::num_simd_work_items(3]] // expected-error
};

where the X, Y, and Z dim expressions in this case here are obtained from Attribute values: A->getXDim(), A->getYDim(), and A->getZDim().

Optionalllvm::APSInt XDimVal = A->getXDimVal(Ctx);
Optionalllvm::APSInt YDimVal = A->getYDimVal(Ctx);
Optionalllvm::APSInt ZDimVal = A->getZDimVal(Ctx);

struct TRIFuncObjBad {
[[intel::reqd_work_group_size(4)]]
[[intel::num_simd_work_items(3]] // expected-error
};